pkg/managed/deployment.go

```diff
@@ -73,8 +76,105 @@ func CreateDeployments(mbPair *types.ModelBindingPair, filteredConnections map[s
 	return nil
 }
 
-// Constructs the necessary Deployments for the specified ManagedCluster in the given VerrazzanoBinding
-func newDeployments(binding *v1beta1v8o.VerrazzanoBinding, managedCluster *types.ManagedCluster, manifest *util.Manifest, verrazzanoURI string, sec monitoring.Secrets) ([]*appsv1.Deployment, error) {
+// DeleteDeployments deletes deployments for a given binding.
+func DeleteDeployments(mbPair *types.ModelBindingPair, filteredConnections map[string]*util.ManagedClusterConnection) error {
+	glog.V(6).Infof("Deleting Deployments for VerrazzanoBinding %s", mbPair.Binding.Name)
+
+	// Delete Deployments associated with the given VerrazzanoBinding (based on labels selectors)
+	for clusterName, managedClusterConnection := range filteredConnections {
+		managedClusterConnection.Lock.RLock()
+		defer managedClusterConnection.Lock.RUnlock()
+
+		selector := labels.SelectorFromSet(util.GetManagedBindingLabels(mbPair.Binding, clusterName))
+
+		existingDeploymentList, err := managedClusterConnection.DeploymentLister.List(selector)
+		if err != nil {
+			return err
+		}
+		for _, deployment := range existingDeploymentList {
+			glog.V(4).Infof("Deleting Deployment %s:%s", deployment.Namespace, deployment.Name)
+			err := managedClusterConnection.KubeClient.AppsV1().Deployments(deployment.Namespace).Delete(context.TODO(), deployment.Name, metav1.DeleteOptions{})
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// CleanupOrphanedDeployments deletes deployments that have been orphaned.   Deployments can be orphaned when a binding
+// has been changed to not require a deployment or the deployment was moved to a different cluster.
+func CleanupOrphanedDeployments(mbPair *types.ModelBindingPair, availableManagedClusterConnections map[string]*util.ManagedClusterConnection) error {
+	glog.V(6).Infof("Cleaning up orphaned Deployments for VerrazzanoBinding %s", mbPair.Binding.Name)
+
+	// Get the managed clusters that this binding applies to
+	matchedClusters, err := util.GetManagedClustersForVerrazzanoBinding(mbPair, availableManagedClusterConnections)
+	if err != nil {
+		return nil
+	}
+
+	for clusterName, mc := range mbPair.ManagedClusters {
```

Comment:
Curious if we can combine the 2 loops in (116 and 150) to just loop through all the clusters

Reply:
This code follows the same pattern for other resources.  I think if we can combine the 2 loops we change it for all the resources and do it in a separate Jira.

Reply:
Don't bother, see slack